libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -497,16 +499,23 @@ bool AC_WPNav::advance_wp_target_along_track(float dt)
     if (!_this_leg_is_spline) {
         // update target position, velocity and acceleration
         target_pos = _origin;
-        s_finished = _scurve_this_leg.advance_target_along_track(_scurve_prev_leg, _scurve_next_leg, _wp_radius_cm, _flags.fast_waypoint, _track_scalar_dt * vel_time_scalar * dt, target_pos, target_vel, target_accel);
+        s_finished = _scurve_this_leg.advance_target_along_track(_scurve_prev_leg, _scurve_next_leg, _wp_radius_cm, _flags.fast_waypoint, _track_scalar_dt * vel_scaler_dt * dt, target_pos, target_vel, target_accel);
     } else {
         // splinetarget_vel
         target_vel = curr_target_vel;
-        _spline_this_leg.advance_target_along_track(_track_scalar_dt * vel_time_scalar * dt, target_pos, target_vel);
+        _spline_this_leg.advance_target_along_track(_track_scalar_dt * vel_scaler_dt * dt, target_pos, target_vel);
         s_finished = _spline_this_leg.reached_destination();
     }
 
-    target_vel *= vel_time_scalar;
-    target_accel *= sq(vel_time_scalar);
+    Vector3f accel_offset;
+    if (is_positive(target_vel.length_squared())) {
+        Vector3f track_direction = target_vel.normalized();
+        accel_offset = track_direction * _offset_accel * target_vel.length() / _wp_desired_speed_xy_cms;
```

Comment:
Sorry for the naivete. Could you explain this scaling?
`* target_vel.length() / _wp_desired_speed_xy_cms`

Reply:
@lthall I think, this one is for you, sir.

Reply:
when everything is moving along without restriction, ie no time compression.
_offset_vel = _wp_desired_speed_xy_cms

we change _offset_vel assuming that the aircraft is moving at a constant velocity of  _wp_desired_speed_xy_cms, generating _offset_accel to move _offset_vel up and down. We then use _offset_vel / _wp_desired_speed_xy_cms to generate the time extension ratio.

Now we are not always trying to travel at _wp_desired_speed_xy_cms so the real tangential acceleration that we need to apply to speed up and slow down based on _offset_vel / _wp_desired_speed_xy_cms becomes:
accel_offset = track_direction * _offset_accel * target_vel.length() / _wp_desired_speed_xy_cms;